src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/ExpressionParser.cs

```diff
@@ -265,20 +265,31 @@ private Expression ParseChainedExpression(string resourceType, SearchParameterIn
 
         private Expression ParseSearchValueExpression(SearchParameterInfo searchParameter, string modifier, string value)
         {
-            SearchModifierCode? parsedModifier = ParseSearchParamModifier();
+            SearchModifierCode? parsedModifier = null;
+            string targetTypeModifier = null;
+            if (!string.IsNullOrEmpty(modifier))
+            {
+                parsedModifier = ParseSearchParamModifier();
+
+                if (parsedModifier == null)
+                {
+                    targetTypeModifier = modifier;
+                }
+            }
 
-            return _searchParameterExpressionParser.Parse(searchParameter, parsedModifier, value);
+            return _searchParameterExpressionParser.Parse(searchParameter, parsedModifier, targetTypeModifier, value);
 
             SearchModifierCode? ParseSearchParamModifier()
             {
-                if (string.IsNullOrEmpty(modifier))
+                if (SearchParamModifierMapping.TryGetValue(modifier, out SearchModifierCode searchModifierCode))
                 {
-                    return null;
+                    return searchModifierCode;
                 }
 
-                if (SearchParamModifierMapping.TryGetValue(modifier, out SearchModifierCode searchModifierCode))
+                // Modifier on a Reference Search Parameter can be used to restrict target type
+                if (searchParameter.Type == SearchParamType.Reference && searchParameter.TargetResourceTypes.Contains(modifier, StringComparer.OrdinalIgnoreCase))
                 {
-                    return searchModifierCode;
+                    return null;
```

Comment:
Should we return `SearchModifierCode.Type` here? It seems to match this the [description](http://hl7.org/fhir/R4/codesystem-search-modifier-code.html) 
> The search parameter only applies to the Resource Type specified as a modifier (e.g. the modifier is not actually :type, but :Patient etc.).